File: src/20.有效的括号.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let map = new Map([
        [')', '('],
        [']', '['],
        ['}', '{'],
    ])
    let queue = []
    for (let i = 0 ; i < s.length; i++) {
        if (queue.length && queue[queue.length-1] === map.get(s.charAt(i))) {
            queue.pop()
        }
        else {
            queue.push(s.charAt(i))
        }
    }
    return queue.length === 0
};
// @lc code=end

File: src/20.有效的括号.rs
/*
 * @lc app=leetcode.cn id=20 lang=rust
 *
 * [20] 有效的括号
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut s: Vec<char> = s.chars().collect();
        let mut stack: Vec<char> = vec![];
        let map = HashMap::from([
            ('(', ')'),
            ('[', ']'),
            ('{', '}'),
        ]);
        for i in s {
            if stack.len() > 0 && map.get(&stack.last().unwrap()) == Some(&i) {
                stack.pop();
            } else {
                stack.push(i);
            }
        }
        // println!("{:?}", stack);

        stack.len() == 0
    }
}
// @lc code=end

